File: app/src/main/java/com/dougkeen/bart/backend/TransitFeedClient.java
package com.dougkeen.bart.backend;

import java.io.IOException;

/** Fetches the complete realtime input used by the transit backend. */
public interface TransitFeedClient {
    TransitFeedSnapshot fetch() throws IOException;

    /**
     * Fetches the two realtime feeds independently. Implementations that
     * already have a combined endpoint may keep the compatibility default;
     * network clients should override this to preserve partial successes.
     */
    default TransitFeedFetchResult fetchFeeds() {
        try {
            return TransitFeedFetchResult.complete(fetch());
        } catch (Exception exception) {
            return TransitFeedFetchResult.failed(exception);
        }
    }
}

File: app/src/main/kotlin/in/izyum/bart/backend/TransitFeedClient.kt
package `in`.izyum.bart.backend

import com.google.transit.realtime.GtfsRealtime
import java.io.IOException

/** Fetches the independent realtime feeds used by the transit backend. */
interface TransitFeedClient {
    @Throws(IOException::class)
    fun fetchFeeds(): TransitFeedFetchResult

    /** Fetches only trip updates for latency-sensitive departure refreshes. */
    @Throws(IOException::class)
    fun fetchTripUpdates(): GtfsRealtime.FeedMessage =
        fetchFeeds().tripUpdates
            ?: throw IOException("Trip updates were unavailable")
}
